lib/rules/indent-function-args.js

```diff
@@ -0,0 +1,150 @@
+/**
+ * @fileoverview Check for function args proper indentation.
+ * @author uberVU
+ * @copyright 2015 uberVU. All rights reserved.
+ * See LICENSE file in root directory for full license.
+ */
+'use strict';
+var _ = require('lodash');
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = function(context) {
+
+  // variables should be defined here
+  var indentMode = 0;
+
+  //--------------------------------------------------------------------------
+  // Helpers
+  //--------------------------------------------------------------------------
+
+  function getNodeIndent(node, byLastLine, excludeCommas, indentType) {
+    /**
+    * Get node indent
+    *
+    * @param {ASTNode|Token} node Node to examine
+    * @param {boolean} [byLastLine=false] get indent of node's last line
+    * @param {boolean} [excludeCommas=false] skip comma on start of line
+    * @returns {int} Indent
+    */
+    var token = byLastLine ? context.getLastToken(node)
+                           : context.getFirstToken(node),
+        src = context.getSource(token, token.loc.start.column),
+        skip = excludeCommas ? ',' : '',
+        regExp;
+
+    if (indentType === 'space') {
+      regExp = new RegExp('^[ ' + skip + ']+');
+    } else {
+      regExp = new RegExp('^[\t' + skip + ']+');
+    }
+
+    var indent = regExp.exec(src);
+    return indent ? indent[0].length : 0;
+  }
+
+  function reportError(node, context) {
+    /**
+    * Reports error for node and indentMode
+    *
+    * @param {ASTNode|Token} node Node to examine
+    * @param {Object} context Rule context object.
+    */
+
+    switch (indentMode) {
+      case 'spaces':
+      case 'line':
+        context.report({
+          node: node,
+          message: 'Indent at 4 spaces.'
+        });
+        break;
+      case 'paren':
+        context.report({
+          node: node,
+          message: 'Indent at parent parenthesis.'
+        });
+        break;
+      default:
+        context.report({
+          node: node,
+          message: 'Indent at 4 spaces or paren.'
+        });
+    }
+  }
+
+  function checkIndent(mode) {
+    /**
+    * Checks the prev indentation (indentMode) against the current one (mode)
+    *
+    * @param {String} mode Current indentation mode.
+    *     Can be 'line', 'paren', 'spaces'.
+    * @param {String} indentMode The indentation that should be enforced across
+    *     all other arguments. Can be 'line', 'paren', 'spaces' and by default
+    *     0 (unset).
+    *
+    * @return {Boolean} Returns true if argument is ok indented or false if not.
+    */
+    switch (indentMode) {
```

Comment:
This function is very hard to figure out + `switch` statements are a code smell (adding a new case means modifying an already working class).

What I would do is use the [chain-of-responsibility pattern](https://en.wikipedia.org/wiki/Chain-of-responsibility_pattern):

Define a list of possible indentation modes (X spaces from the call, start of paren etc) and feed the current indentation through each of them until one of them approves it. If no one approves it then you throw an error stating that you must indent in at least one of the modes (each one could also define a description).

You only need to ask them for the first line, once you get a match you store it and you use it for the next lines. Or you keep trying the lines until you find a match and then you use that — for cases where the first line isn't correctly matched.

```js
var modes = [
  at4spacesFromFunctionCall,
  alignedWithParen,
  ...
];

requiredIndentation = null

for each node
  if line changed
    newIndentation = indentation(node)
    
    // Now we decide how all the lines will be indented.
    if not requiredIndentation
      requiredIndentation = getIndentationMode(node)

      if not requiredIndentation
        log error
        continue // we try to find another correctly indented line

    // This line must be aligned in the same way as the previous line.
    if newIndentation is not requiredIndentation
      log error

function getIndentationMode(node) {
  for mode in modes
    if mode.accepts(node)
      return indentation(node)

  return null
}

// Example of indentation mode.
alignedWithParen = {
  /**
   * @returns {Boolean}
   */
  accepts: function(node, parent) {
    // figure out if node is aligned with the paren
  }
}
```